temporalio/worker/_workflow.py

```diff
@@ -193,6 +206,59 @@ async def _handle_activation(
     ) -> None:
         global LOG_PROTOS
 
+        def get_traceback_for_thread(
+            thread_id: int,
+        ) -> Tuple[Optional[TracebackType], bool]:
+            """Take a thread id and attempt to construct a traceback from its frames.
+
+            Args:
+                thread_id: Thread ID to be checked.
+
+            Returns:
+                A tuple containing:
+                    0: A TracebackType if the supplied thread id is valid, None otherwise.
+                    1: A bool that determines whether the error message should warn about a possibly incomplete traceback.
+            """
+            # retrieve all frames
+            frame = sys._current_frames().get(thread_id)
+
+            # if the id is not present, the thread must have terminated or there is another error
+            if not frame:
+                return (None, False)
+
+            # Construct tb stack, might be a better way/builtin to do this
+
+            # gather all frames
+            tb_frames = [frame]
+            while frame.f_back:
+                frame = frame.f_back
+                tb_frames.append(frame)
+
+            top_tb = TracebackType(
+                None,
+                tb_frame=tb_frames[0],
+                tb_lasti=tb_frames[0].f_lasti,
+                tb_lineno=tb_frames[0].f_lineno,
+            )
+            tb_frames.reverse()
+            tb_frames.pop()
+
+            tb = top_tb
+            for f in tb_frames:
```

Comment:
Suppose the workflow is timing out because two coroutines are playing some sort of infinite ping-pong (e.g. the workflow in https://github.com/temporalio/sdk-python/issues/527), will the stack trace be huge, and so we need to limit the number of frames? Or does that asyncio coroutine switching not contribute any stack frames?

Reply:
We probably should set a size limit on this traceback. There is a server-side truncation for failure stack traces that are too large, but it's technically possible we could exceed the 4MB limit to even send the message. I'd truncate at something extreme like 200000 bytes (i.e. 200kb) maybe?